File: apps/daemon/internal/intents/dispatch/undo/dispatch.go
package undo

import (
	"fmt"

	"vocoding.net/vocode/v2/apps/daemon/internal/intents"
	protocol "vocoding.net/vocode/v2/packages/protocol/go"
)

// DispatchUndo maps a planner undo intent to the wire payload for the extension (host applies).
func DispatchUndo(u intents.UndoIntent) (protocol.UndoDirective, error) {
	if err := intents.ValidateUndoIntent(u); err != nil {
		return protocol.UndoDirective{}, fmt.Errorf("undo intent: %w", err)
	}
	return protocol.UndoDirective{Scope: string(u.Scope)}, nil
}

File: apps/vscode-extension/src/directives/undo/dispatch.ts
import type { UndoDirective } from "@vocode/protocol";

import { applyUndoDirective } from "./transcript-undo-ledger";

/** Applies one undo directive (host undo stack / transcript ledger). */
export function dispatchUndo(
  directive: UndoDirective | undefined,
): Promise<boolean> {
  return applyUndoDirective(directive);
}
